### PR TITLE
Route requests and route replies

### PR DESCRIPTION
**Pull request by puddly**
This PR aims to integrate routing and route discovery tables into ziggurat. More precisely:

1. We no longer allow single-hop route-less sending: routes to devices need to be calculated.
2. To do that, we send route requests.
3. And process route replies, both expected an unsolicited.

A future PR will properly age routes, propagate routing errors, and hopefully stabilize network management into a state where ziggurat can be properly tested on a large production network.

### CODE REVIEW

File: crates/stack/src/zigbee_stack.rs [hunk 1]
```diff
@@ -36,13 +37,20 @@ mod route;
...
-    ((1.0 - (lqi as f32) / 255.0) * 6.0 + 1.0).round() as u8
+    match lqi {
+        0..=16 => 7,
+        17..=32 => 6,
+        33..=64 => 5,
+        65..=96 => 4,
+        97..=128 => 3,
+        129..=192 => 2,
+        193..=255 => 1,
+        // 0 corresponds to "unknown LQI"
```
Comment (yara-blue) [inline crates/stack/src/zigbee_stack.rs:51]: this could return an option instead when the LQI is zero. We could also make the return type a newtype: `LinkCost` to enhance readability further.

Comment (puddly) [reply crates/stack/src/zigbee_stack.rs:51]: Any valid LQI will map to a link cost of 1-7 (LQI of 0 maps to a link cost of 7), it's just that the spec uses `0` to mean "unknown" link costs. I think we can deviate from the spec here and make it an `Option` as you said.

File: crates/stack/src/zigbee_stack.rs [hunk 2]
```diff
@@ -747,7 +791,14 @@ impl ZigbeeStack {
         drop(state);
 
         log::info!("Decrypted frame: {decrypted_nwk_frame:#?}");
-        self.handle_decrypted_frame(&decrypted_nwk_frame, lqi, rssi);
+
+        // TODO: all 802.15.4 frames should be coming in with 16 bit addressing, right?
```
Comment (yara-blue) [inline crates/stack/src/zigbee_stack.rs:805]: Not sure, but we could just panic in the none case by unwrapping the `src_address`. We will find out quite quickly that way.

File: crates/stack/src/zigbee_stack.rs [hunk 3]
```diff
@@ -1017,9 +1085,186 @@ impl ZigbeeStack {
...
             return;
         }
+
+        // Both of these fields SHALL be set according to the spec
+        if route_reply_cmd.responder_eui64.is_none() || route_reply_cmd.originator_eui64.is_none() {
+            return;
+        }
+
+        let sender_nwk = nwk_frame.nwk_header.source;
+        let rrep_id = route_reply_cmd.route_request_identifier;
```
Comment (yara-blue) [inline crates/stack/src/zigbee_stack.rs]: might make more sense to leave these variables out, they make the function a bit longer (and its already really long). I can also take a stab at refactoring this later.

File: crates/stack/src/zigbee_stack.rs [hunk 4]
```diff
@@ -1017,9 +1085,186 @@ impl ZigbeeStack {
...
+            log::debug!("Ignoring route reply from neighbor with zero outgoing cost");
+            return;
+        }
+
+        let updated_path_cost = rrep_path_cost.saturating_add(neighbor.outgoing_cost);
+        let route_discovery_table_key = (rrep_originator, rrep_id);
+
+        // TODO: clean this up, the mutability problems caused by the shared state mutex
+        // make having two mutable borrows at once very difficult
+        let discovery_entry = match state
```
Comment (yara-blue) [inline crates/stack/src/zigbee_stack.rs]: `let else` might make this nicer.

```rust
let Some(entry) = state.things.things else {
  log then return
}
```

Its a pattern I see a lot (log if none and return). We could do that using errors in the future? That would look like:

```rust
let entry = state.nib.nwk_etc.get().ok_or(Error::UnknownRouteDiscovery)?;
```

File: crates/stack/src/zigbee_stack.rs [hunk 5]
```diff
@@ -1017,9 +1085,186 @@ impl ZigbeeStack {
...
+            }
+        };
+
+        // If we are the originator, handling is simplified
+        if rrep_originator_eui64 == our_ieee_address {
+            match routing_entry.status {
+                route::Status::Inactive
+                | route::Status::DiscoveryFailed  // Is this correct to have?
+                | route::Status::DiscoveryUnderway => {
+                    log::debug!("Setting routing entry for NWK {:?} to active, with next hop {:?} (residual cost {})",
```
Comment (yara-blue) [inline crates/stack/src/zigbee_stack.rs:1191]: maybe extract the body to a function? (same for the other branch)

File: crates/stack/src/zigbee_stack.rs [hunk 6]
```diff
@@ -1071,55 +1333,58 @@ impl ZigbeeStack {
...
-                key_sequence_number: state.nib.nwk_active_key_seq_number,
-            }),
+            aux_header: None, // will be replaced
             payload: NwkRouteReplyCommand {
                 multicast: false,
                 route_request_identifier: route_request_cmd.route_request_identifier,
                 originator_nwk: nwk_frame.nwk_header.source,
                 responder_nwk: state.nib.nwk_network_address,
-                path_cost: 1, // :)
+                path_cost: path_cost,
```
Comment (yara-blue) [inline crates/stack/src/zigbee_stack.rs]: should just be: `path_cost,`

off-topic: do you have rust-analyzer (the LSP) set up? might help with stuff like this. Let me know if you need any help getting it working.

**Review by yara-blue — COMMENTED** [review]
Trying to help by reviewing where possible. Let me know if its more of a chore then help. I can imagine you are trying to get things working asap and plan to refactor after, in which case my comments might be more distracting then useful.

File: crates/stack/src/zigbee_stack.rs (hunk 4 above)
Comment (puddly) [reply crates/stack/src/zigbee_stack.rs]: Both look excellent, the error handling pattern specifically. I'm going to try and run through the current error strings later and turn them into enums to permit this, they're getting a little unwieldy already.

Comment (yara-blue) [reply crates/stack/src/zigbee_stack.rs]: >  I'm going to try and run through the current error strings later and turn them into enums to permit this

We can divide that work, no reason for you to do all that on your own.

I would recommend using [thiserror](https://crates.io/crates/thiserror), it has become kind of the standard for error handling within libs. It does a few nice things for you, it generates a [Error](https://doc.rust-lang.org/nightly/core/error/trait.Error.html) trait implementation for you. If you annotate the errors with `#[source]` or `#[from]` the library user gets an "error backtrace", as in:

totally made up:
```
- Error sending packet
- Could not encode
- Sequence number incorrect
```
which is printed from something like this:
```rust
Error::Sending(SendError::Encode(EncodeError::SequenceNumber))
```

**Comment by puddly** [comment]
Not at all, I appreciate the help!

Right now I'm trying to get this branch into a functional state, since route discovery works 50% of the time (and everything is very snappy when it does) but seems to fail the other 50% due to timeouts (or possibly mutex contention, I may look at tokio RwLocks as a stopgap).

Route reply re-broadcasting is also something that needs to be finished up but it's not necessary to function as a coordinator on a network, it just helps devices route through us.

Once that's done I think everything needs a proper refactor, especially since there are strange gaps in both the spec and the observed behavior of other devices when it comes to routing, which will let us ditch these unwieldy tables.

**Comment by yara-blue** [comment]
> Right now I'm trying to get this branch into a functional state, since route discovery works 50% of the time (and everything is very snappy when it does) but seems to fail the other 50% due to timeouts (or possibly mutex contention, I may look at tokio RwLocks as a stopgap).

If it happens in async code you could try [tokio-console](https://github.com/tokio-rs/console). There is some work to setting it up though but it should be manageable. Another alternative is to overhaul the architecture and remove the need for shared state (and thus locks). I have some ideas for that if your interested, though I can imagine you just want to get this reliably working first before overhauling it.

**Comment by puddly** [comment]
I think this ad-hoc routing implementation is practically done. I've tested it with Home Assistant after adding ziggurat as a radio library and things work as expected on my small test network.

Route requests and route replies are functional, though relayed route requests are still untested. Conceptually, they work and help populate routing information using all available information.

---

Once this PR is merged, the fast development is coming to a close and I think it'll be time to heavily refactor everything. There are still a few missing pieces but they aren't critical for functionality and testing, just edge cases and research opportunities:

1. Internal retrying: we should actually be retrying in the stack three times, with jittered 50ms delays between each retry. There are a few other spots with retries/jittering.
2. Error status broadcasts: routing failures should result in short broadcasts that notify nearby nodes of the routing failure.
3. Broadcast relaying: incoming broadcasts should be relayed with a decremented radius. There's some semi-complex logic in the spec for determining when to stop rebroadcasting by checking to see if every known neighbor router has relayed the frame.
4. Device joining: we don't allow nodes to join the network and assume you've joined them to your network already. This should be relatively straightforward.
5. Multicast route request handling: I'm not too sure how this works, will need to investigate.
6. Source routing: this is very useful and will be the routing implementation we will eventually switch to, since it will allow for the coordinator to decide the exact path of a packet (as opposed to hoping that the next-hop router does what it says).
7. Stochastic routing: I'd like to do some more research but I've seen that existing Zigbee routing implementations just stick to a single route and wait for it to stop working before finding a new one. This is too simple, IMO, and creates choke points in the network topology, reducing effective concurrency and adding unnecessary delays when routes inevitably break. I think we can do better by adding some probabilities to the mix and rerouting concurrent requests via alternate paths.

---

Unrelated to the above, but I've been re-reading this single page of the spec for a very, very long time and have come to realize that English really doesn't make for a good programming language 😅.

image

**Comment by yara-blue** [comment]
> I think this ad-hoc routing implementation is practically done. I've tested it with Home Assistant after adding ziggurat as a radio library and things work as expected on my small test network.

Wonderful news :partying_face: ! I got an official esp32-c6 devkit from the espressif people, that should run open-thread (and thus spinal). I'll be able to start testing with that. I'll also write up my steps to get from a raw c6 to an open-thread/spinal + ziggurat system. Its a pretty popular and easy to source chip/dev board, so that might help others.

The c6 is kind of a beast (performance wise) so it might even be a good candidate for a future no-std pure rust coordinator without spinal. Thinking about it though, running the coordinator on the host system is still the best approach if you ask me. It makes updating trivial for the end user. So that was a great idea.

> Once this PR is merged, the fast development is coming to a close and I think it'll be time to heavily refactor everything.

Looking forward to it, once this is merged I'll make an issue listing any major restructuring opportunities I can find. Then once we agree we can work on those together.

> There are still a few missing pieces but they aren't critical for functionality and testing, just edge cases and research opportunities:

You have some great ideas there, should we track them in a separate issue so they do not get buried?

> Unrelated to the above, but I've been re-reading this single page of the spec for a very, very long time and have come to realize that English really doesn't make for a good programming language 😅.

Oh that is awful to read (thanks for showing it however!), they should really have gone for a flowchart (or just pseudo-code).

**Comment by puddly** [comment]
I've been running my 60 device home network with Ziggurat for about an hour now with no major issues to report, I think this PR is ready to merge.

Some small bugs:

1. We don't implement NWK-level retries at the moment. If a request fails then it'll need to be retried by zigpy 5s later, which is visibly slow.
2. There is some small bit of routing information that we are missing over-the-air. For some devices route discovery works well, for others we have to wait to receive a packet from the device to build a provisional path back to it.
